src/components/forms/CharacterCount/CharacterCount.tsx

```diff
@@ -0,0 +1,165 @@
+import React, { useState } from 'react'
+import classnames from 'classnames'
+
+import { TextInput, TextInputRef } from '../TextInput/TextInput'
+import { Textarea, TextareaRef } from '../Textarea/Textarea'
+import { FormGroup } from '../FormGroup/FormGroup'
+
+interface CharacterCountProps {
+  id: string
+  name: string
+  maxLength: number
+  className?: string
+  label?: React.ReactNode
+  isTextArea?: boolean
+  textRef?: // currently getting an error when adding this as ref
+  | string
+    | ((instance: HTMLInputElement | null) => void)
+    | ((instance: HTMLTextAreaElement | null) => void)
+    | React.RefObject<HTMLInputElement>
+    | React.RefObject<HTMLTextAreaElement>
+    | null
+    | undefined
+}
+
+export const CharacterCount = (
+  props: CharacterCountProps &
+    React.InputHTMLAttributes<HTMLInputElement> &
+    React.TextareaHTMLAttributes<HTMLTextAreaElement>
+): React.ReactElement => {
+  const {
+    id,
+    name,
+    className,
+    label,
+    maxLength,
+    defaultValue,
+    isTextArea,
+    textRef,
+    type, // Conflicts with type attribute already set on TextInput component
+    ...inputProps
+  } = props
+
+  const defaultLength =
+    defaultValue === undefined ? 0 : defaultValue.toString().length
+  const maxNum = maxLength === undefined ? 0 : maxLength
+
+  /* Ideally defined as i18n translation strings */
+  const emptyMessageFormat = `${maxLength} characters allowed`
```

Comment:
I think we agreed to pass lang strings as props until we have a bigger discussion. Suggestion - the lang could be a less props if we ignore singular v. plural for now- `emptyText`, `remainingText`, `overLimitText`. We can write the english defaults as "character(s) over limit" for now.

Reply:
thinking long term for components like this that render complex strings (of which I doubt there will be many in this lib), this is where I think it would be more useful to expose a function prop that gets called with the current `len` value and returns the whole message string, so the user can have complete control over the logic and the copy. and the default value of this prop can be the existing `limitMessage` fn with the default English strings encapsulated in the fn.

Reply:
@suzubara did you see this comment? https://github.com/trussworks/react-uswds/pull/217#pullrequestreview-423042971 seems like we are on the same page.  Can we just do that now?  I think that's more useful.

Reply:
ohh got it. So, yes I think we agree, my only suggestion is that there be two props, one that returns the message copy (ie, `message`), and one that indicates validity (ie, `isValid` or `hasError`) (see my note about disable native validation). It is very likely that a user who is doing a custom implementation for copy will use the same code on their end to feed both props, but I think the component API will be easier to work with if they are decoupled. To me, a prop called `validateLength` sounds like an imperative function that might trigger validation, when really it should just return the message string.